braid/residual.c

```diff
@@ -43,19 +44,45 @@ _braid_Residual(braid_Core        core,
    braid_Int        ilower   = _braid_GridElt(grids[level], ilower);
    braid_Real      *ta       = _braid_GridElt(grids[level], ta);
 
+   /* this decides whether Delta correction needs to be computed on this level,
+   *  which depends on from where the residual is called
+   */
+   braid_Int delta_correct;
+   delta_correct = _braid_CoreElt(core, delta_correct)
+                   &&  ( iter  >= _braid_CoreElt(core, delta_defer_iter) )
+                   &&  ( level >= _braid_CoreElt(core, delta_defer_lvl) )
+                   && !( calling_function == braid_ASCaller_Residual
+                      && level == _braid_CoreElt(core, delta_defer_lvl) );
+
    braid_BaseVector rstop;
    braid_Int        ii;
 
    ii = index-ilower;
-   _braid_StepStatusInit(ta[ii-1], ta[ii], index-1, tol, iter, level, nrefine, gupper, status);
+   
+   if ( delta_correct )
+   {  /* Give the user access to the basis vectors through StepStatusGetBasisVec */
+      _braid_StepStatusInit(ta[ii-1], ta[ii], index-1, tol, iter, level, nrefine, gupper, calling_function, r->basis, status);
+
+      /* By default: r = ustop - \Phi(ustart)*/
+      _braid_GetUInit(core, level, index, r, &rstop);
+      _braid_BaseStep(core, app, rstop, NULL, r, level, status);
+      _braid_CoreFcn(core, sum)(app, 1.0, ustop->userVector, -1.0, r->userVector);
+      _braid_BaseSumBasis(core, app, 0., r->basis, -1.0, r->basis);
+
+      return _braid_error_flag;
+   }
+   /* else, default behavior */
```

Comment:
do we want a formal else statement here, and then a global return _braid_error_flag at the end of the routine?